Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/TargetController.cs

```diff
@@ -0,0 +1,143 @@
+﻿using UnityEngine;
+using Random = UnityEngine.Random;
+using Unity.MLAgents;
+using UnityEngine.Events;
+
+namespace Unity.MLAgentsExamples
+{
+    /// <summary>
+    /// Utility class to allow target placement and collision detection with an agent
+    /// Add this script to the target you want the agent to touch.
+    /// Callbacks will be triggered any time the target is touched with a collider tagged as 'tagToDetect'
+    /// </summary>
+    public class TargetController : MonoBehaviour
+    {
+        
+        [Header("Collider Tag To Detect")]
+        public string tagToDetect = "agent"; //collider tag to detect 
+        
+        [Header("Target Placement")]
+        public float spawnRadius; //The radius in which a target can be randomly spawned.
+        public bool respawnIfTouched; //Should the target respawn to a different position when touched
+        
+        [Header("Target Fell Protection")]
+        public bool respawnIfFallsOffPlatform = true; //If the target falls off the platform, reset the position.
+        public float fallDistance = 5; //distance below the starting height that will trigger a respawn 
+
+
+        private Vector3 m_startingPos; //the starting position of the target
+        private Agent m_agentTouching; //the agent currently touching the target
+
+        [System.Serializable]
+        public class TriggerEvent : UnityEvent<Collider>
+        {
+        }
+
+        [Header("Trigger Callbacks")] public bool triggerIsTouching;
+        public TriggerEvent onTriggerEnterEvent = new TriggerEvent();
+        public TriggerEvent onTriggerStayEvent = new TriggerEvent();
+        public TriggerEvent onTriggerExitEvent = new TriggerEvent();
+
+        [System.Serializable]
+        public class CollisionEvent : UnityEvent<Collision>
+        {
+        }
+
+        [Header("Collision Callbacks")] public bool colliderIsTouching;
+        public CollisionEvent onCollisionEnterEvent = new CollisionEvent();
+        public CollisionEvent onCollisionStayEvent = new CollisionEvent();
+        public CollisionEvent onCollisionExitEvent = new CollisionEvent();
+
+        // Start is called before the first frame update
+        void OnEnable()
+        {
+            m_startingPos = transform.position;
+            if (respawnIfTouched)
+            {
+                MoveTargetToRandomPosition();
+            }
+        }
+
+        void Update()
+        {
+            if (respawnIfFallsOffPlatform)
+            {
+                if (transform.position.y < m_startingPos.y - fallDistance)
+                {
+                    Debug.LogError($"{transform.name} Fell Off Platform");
```

Comment:
Do we want to log this as an error?

Reply:
no i don't think so i changed it to a Log()